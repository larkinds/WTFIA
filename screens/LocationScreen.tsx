import MapView from 'react-native-maps';
import { useContext, useState } from 'react';
import { StyleSheet, Text, View, ActivityIndicator } from 'react-native';
import Countdown from '../components/Countdown';
import { LocationContext } from '../context/LocationContext';
import RefreshButton from '../components/RefreshButton';

export default function LocationScreen( ) {
  const {neighborhood, region, fetchLocation} = useContext(LocationContext);
  const [reloading, setReloading] = useState(false);

  async function refreshLocation() {
    setReloading(true);
    fetchLocation();
    
    setTimeout(() => {
      setReloading(false);
    }, 11100)
  }

  return (
    <View style={styles.container}>
          <MapView
            style={StyleSheet.absoluteFillObject}
            mapType="mutedStandard"
            showsUserLocation={true}
            region={region}
          />
          <View>
            {neighborhood === '' ? (
              reloading ? (<ActivityIndicator style={styles.activityIndicator} size="large"  color="black"/>) : (<Countdown />)
            ) : (
              <>
              <View style={styles.locationVisibleView}>
                <Text style={{color: "black", fontSize: 20, marginLeft:5, marginRight: 5}}>{neighborhood} <RefreshButton onPressFunction={refreshLocation} /></Text>
              </View>
              </>
            )}
          </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    width: '100%',
    height: '100%',
  },
  // locationsDisabledView: {
  //   display: 'flex',
  //   alignItems: 'center',
  // },
  // retryPermissionsView: {
  //   flex: 1,
  //   fontSize: 14,
  //   margin: 4,
  // },
  activityIndicator: {
    backgroundColor: "#FDF1D9", 
    top: 430, 
    padding: 10
  },
  locationVisibleView: {
    backgroundColor: "#E1E0DB",
    top: 200,
    padding: 2,
    paddingBottom: 8
  }
});
