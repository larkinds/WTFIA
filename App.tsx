import React, { useState, useEffect } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import * as Location from 'expo-location';
import { LocationObject } from 'expo-location';
import Geocoder from 'react-native-geocoding';
import { GeographicInfo, LocationContext } from './context/LocationContext';
import LocationScreen from './screens/LocationScreen';
import SearchScreen from './screens/SearchScreen'
import HomeScreen from './screens/HomeScreen';

const Stack = createNativeStackNavigator();

export default function App() {
  const [errorMessage, setErrorMessage] = useState<string | null>(null);
  const [currentLocation, setCurrentLocation] = useState<LocationObject | GeographicInfo | null>(
    null
  );
  const [neighborhood, setNeighborhood] = useState<string>('');
  let latitude = Math.floor(Math.random() * (90 - -90 + 1) + -90);
  let longitude = Math.floor(Math.random() * (180 - -180 + 1) + -180);
  const [region, setRegion] = useState({
    latitude,
    longitude,
    latitudeDelta: 130,
    longitudeDelta: 130,
  });

  async function fetchLocation() {
    setNeighborhood("");
    let { status } = await Location.requestForegroundPermissionsAsync();
    if (status !== 'granted') {
      setErrorMessage(
        'You can still use WTFIA by entering your address manually on the Explore page.'
      );
      return;
    };
    let location: LocationObject = await Location.getCurrentPositionAsync();
 
    setCurrentLocation(location);
  };


  useEffect(() => {
    (async () => {
      if (currentLocation && "neighborhood" in currentLocation) {
        setNeighborhood(currentLocation.neighborhood);
        setRegion(currentLocation.region);
        return
      }

      let latitude: number;
      let longitude: number;

      if (currentLocation && "coords" in currentLocation && currentLocation.coords && currentLocation.coords.latitude && currentLocation.coords.longitude) {
        latitude = currentLocation?.coords.latitude;
        longitude = currentLocation?.coords.longitude;
        let tempRegion = {
          latitude,
          longitude,
          latitudeDelta: 0.05,
          longitudeDelta: 0.05,
        };

        setRegion(tempRegion);

        const geocoderKey: string = process.env.EXPO_PUBLIC_GEOCODER_KEY || "";

        Geocoder.init(geocoderKey);

        Geocoder.from(latitude, longitude)
          .then((json) => {
            const addressComponent = json.results[0].address_components;

            const neighborhood = addressComponent.find((component) =>
              component.types.includes('neighborhood') || component.types.includes("locality")
            );

            if (neighborhood) {
              const neighborhoodName = neighborhood.long_name;
              setNeighborhood(neighborhoodName);
            } else {
              console.log(
                'Neighborhood information not found in the response.'
              );
            }
          })
          .catch((error) => {
            console.error('Error:', error);
          });
      }
    })();
  }, [currentLocation]);

  return (
    <LocationContext.Provider value={{ neighborhood, region, setCurrentLocation, fetchLocation }}>
      <NavigationContainer>
        <Stack.Navigator screenOptions={{headerTransparent: true}}>
          <Stack.Screen name="Home" component={HomeScreen} options={{title: ""}} />
          <Stack.Screen name="Location" component={LocationScreen} />
          <Stack.Screen name="Search" component={SearchScreen} />
        </Stack.Navigator>
      </NavigationContainer>
    </LocationContext.Provider>
  );
}
