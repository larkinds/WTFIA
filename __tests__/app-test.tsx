
/**
 * @jest-environment jsdom
 */

import renderer from 'react-test-renderer';
import "jest-canvas-mock";
import App from "../App"
import LocationScreen from '../screens/LocationScreen';


global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}))

test('renders correctly', () => {
  const view = renderer.create(<App />).toJSON();
  expect(view).toMatchSnapshot();
});