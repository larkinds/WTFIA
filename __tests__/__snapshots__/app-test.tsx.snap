// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "black",
                "display": "flex",
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 20,
                  "top": 10,
                }
              }
            >
              ?
            </Text>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 25,
                  "top": 20,
                }
              }
            >
              ?
            </Text>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 32,
                  "top": 2,
                }
              }
            >
              ?
            </Text>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 32,
                  "top": 25,
                }
              }
            >
              ?
            </Text>
            <View
              style={
                {
                  "height": 600,
                  "width": 600,
                }
              }
            >
              <div
                style={
                  {
                    "height": "100%",
                    "overflow": "hidden",
                    "pointerEvents": "auto",
                    "position": "relative",
                    "width": "100%",
                  }
                }
              >
                <div
                  style={
                    {
                      "height": "100%",
                      "width": "100%",
                    }
                  }
                >
                  <canvas
                    style={
                      {
                        "display": "block",
                      }
                    }
                  />
                </div>
              </div>
            </View>
            <View
              style={
                {
                  "backgroundColor": "black",
                  "borderColor": "white",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "fontSize": 20,
                  "marginTop": -60,
                  "padding": 2,
                  "shadowColor": "white",
                  "shadowOffset": {
                    "height": 1,
                    "width": 1,
                  },
                  "shadowOpacity": 1,
                  "shadowRadius": 2,
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    [
                      {},
                    ]
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#007AFF",
                          "fontSize": 18,
                          "margin": 8,
                          "textAlign": "center",
                        },
                        {
                          "color": "white",
                        },
                      ]
                    }
                  >
                    Where the Fuck Am I?
                  </Text>
                </View>
              </View>
            </View>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 40,
                  "top": 20,
                }
              }
            >
              ?
            </Text>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 20,
                  "top": 10,
                }
              }
            >
              ?
            </Text>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 20,
                  "top": 10,
                }
              }
            >
              ?
            </Text>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 25,
                  "top": 7,
                }
              }
            >
              ?
            </Text>
            <Text
              collapsable={false}
              style={
                {
                  "color": "rgba(0, 0, 0, 1)",
                  "left": 32,
                  "top": 0,
                }
              }
            >
              ?
            </Text>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="transparent"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        hideShadow={true}
        largeTitleHideShadow={false}
        title=""
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={true}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
